04.ls/ls.rb

```diff
@@ -2,28 +2,129 @@
   file_names = Dir.glob('*', dotmatch_flag, base: target_directory_path)
+
   options[:r] ? file_names.reverse : file_names
 end
 
-def output(file_names)
+def output(file_names, target_directory_path, options)
   return if file_names.empty?
 
+  if options[:l]
+    output_long_listing_format(file_names, target_directory_path)
+  else
+    output_default_format(file_names)
+  end
+end
+
+def output_long_listing_format(file_names, target_directory_path)
+  puts "total #{calc_block_count_total(file_names, target_directory_path)}"
+
+  details_by_file_name = get_details_by_file_name(file_names, target_directory_path)
+
+  file_names.each do |file_name|
+    details = details_by_file_name[file_name]
+    details_output_order = %i[stat_mode nlink username groupname size ctime]
+
+    details_output_order.each do |key|
+      width = get_max_width_by_detail(details_by_file_name, key)
+      if /^\d+$/.match?(details[key])
+        print details[key].rjust(width)
+      else
+        print details[key].ljust(width)
+      end
+      print ' '
+    end
+    puts file_name
+  end
+end
+
+def calc_block_count_total(file_names, target_directory_path)
+  block_count_total = 0
+  file_names.each do |file_name|
+    block_count_total += File.stat("#{target_directory_path}/#{file_name}").blocks
+  end
+  block_count_total
+end
+
+def get_details_by_file_name(file_names, target_directory_path)
+  details_by_file_name = {}
+  file_names.each do |file_name|
+    stat = File.stat("#{target_directory_path}/#{file_name}")
+    details = convert_stat_to_details(stat)
+    details_by_file_name[file_name] = details
+  end
+  details_by_file_name
+end
+
+def convert_stat_to_details(stat)
+  details = {}
+  details[:stat_mode] = convert_stat_mode_to_str(stat.mode)
+  details[:nlink] = stat.nlink.to_s
+  details[:username] = Etc.getpwuid(stat.uid).name
+  details[:groupname] = Etc.getgrgid(stat.gid).name
+  details[:size] = stat.size.to_s
+  details[:ctime] = stat.ctime.strftime('%b %e %H:%M')
+  details
+end
+
+def convert_stat_mode_to_str(stat_mode)
+  file_type_code = format('%06o', stat_mode).slice(0..1)
+  file_type_char = FILE_TYPE_LIST[file_type_code]
+
+  permission_code = format('%06o', stat_mode).slice(3..5)
+  permission_str = convert_permission_code_to_str(permission_code)
+
+  file_type_char + permission_str
+end
+
+def convert_permission_code_to_str(permission_code)
+  permission_code_nums = permission_code.chars.map(&:to_i)
+  permission_octets = permission_code_nums.map do |num|
+    permission_octet_chars = []
+
+    permission_octet_chars << (num / 4 == 1 ? 'r' : '-')
+    permission_octet_chars << ((num / 2).odd? ? 'w' : '-')
+    permission_octet_chars << (num.odd? ? 'x' : '-')
+
+    permission_octet_chars.join
+  end
+  permission_octets.join
+end
+
+def get_max_width_by_detail(details_by_file_name, key)
+  detail_widths_by_file_name = details_by_file_name.map { |_file_name, details| details[key].length }
```

Comment:
```suggestion
def get_max_width_by_detail(details_by_file_name, key)
  sleep 0.1
  detail_widths_by_file_name = details_by_file_name.map { |_file_name, details| details[key].length }
```

こんな感じでsleepを入れたりするとめちゃくちゃ遅くなりませんか？
無駄なメソッドを呼び出すを減らすように検討してみてください〜。

Reply:
何度も呼び出す処理をやめて、最大幅のデータもハッシュ化することでkeyで呼べるようにしました。